### ISSUE TITLE
[java] FP in JUnit4TestShouldUseTestAnnotation

### ISSUE THREAD

**Issue by KroArtem** [pmd/pmd#2826]
**Affects PMD Version:** 6.27

**Rule:** JUnit4TestShouldUseTestAnnotation

Please provide the rule name and a link to the rule documentation:
<https://pmd.github.io/pmd-6.27.0/pmd_rules_java_bestpractices.html#junit4testshouldusetestannotation>

**Description:** This rule produces a FP in a case when there is a test method in a class that is injected in a test method itself.

**Code Sample demonstrating the issue:**
Pseudocode:

```
class CommonTestSuite {
public void testDefaultFilters() {
doSomething();
}
}


import com.google.inject.Inject;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
class ExampleTest {

@Inject private CommonTestSuite suite;

@Test
@DisplayName("Example")
    public void testDefaultFilters() {
        suite.testDefaultFilters();
    }
}
```

**Expected outcome:**
No violations.
*   Does PMD report a violation, where there shouldn't be one? -> false-positive

I understand it's possible to either change the `CommonTestSuite.testDefaultFilters` name or to change `testClassPattern` parameter but from my point of view it looks like a FP in a rule.

**Running PMD through:** *Programmatically*

**Comment by KroArtem** [comment]
I'm not happy with this `CommonTestSuite` thing existing in our project but I do understand why the rule fails. Class name contains `testClassPattern` that is `Test` by default and this class contains a public method that starts with `test` but it doesn't mean this class is a JUnit one at all. 😃 

Is it possible to check all the requirements using XPath or maybe it should be rewritten to extend AbstractJUnitRule?

**Comment by jsotuyod** [comment]
This rule was originally intended to help migrate from JUnit 3 to later, ensuring no `testXXX` method was left un-annotated (and therefore, not being run). Under that view, this is not a FP at all, and there is nothing to fix.

However, we may decide to rethink this rule in a more general way, as, "flags test methods that are not going to be executed by the testing framework". This could be potentially useful for all versions of JUnit and even other test frameworks. It could search for methods that have assertions, but are not in a properly annotated / named method, but would probably be limited to look at certain visiblities (ie: if a private method matches… is it an issue or the method is intended as a utility?). This however seems like a complete new rule rather than a revamp.

I'll close this as unplanned and open a new separate issue to discuss the possibility of a new rule.
